testing/upload_validation.ts

```diff
@@ -355,1385 +365,1409 @@ const setEndpointPostSetupValidator = zodCompleteObject<SetEndpoint>({
-        ([_key, prop]) => prop.type === ValueType.String && prop.codaType === ValueHintType.CodaInternalRichText,
-      );
-      if (internalRichTextPropertyTuple && !isValidUseOfCodaInternalRichText(data.identity?.packId)) {
-        context.addIssue({
-          code: z.ZodIssueCode.custom,
-          path: ['identity', 'properties', internalRichTextPropertyTuple[0]],
-          message: 'Invalid codaType. CodaInternalRichText is not a supported value.',
-        });
-        return;
-      }
-    }),
-);
+        if (internalRichTextPropertyTuple && !isValidUseOfCodaInternalRichText(data.identity?.packId)) {
+          context.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['identity', 'properties', internalRichTextPropertyTuple[0]],
+            message: 'Invalid codaType. CodaInternalRichText is not a supported value.',
+          });
+          return;
+        }
+      }),
+  );
 
-const objectPropertyUnionSchema = z
-  .union([booleanPropertySchema, numberPropertySchema, stringPropertySchema, arrayPropertySchema, genericObjectSchema])
-  .refine((schema: Schema) => {
-    const schemaForAutocomplete = maybeUnwrapArraySchema(schema);
-    const result =
-      !schemaForAutocomplete ||
-      unwrappedSchemaSupportsAutocomplete(schemaForAutocomplete) ||
-      !('autocomplete' in schemaForAutocomplete && schemaForAutocomplete.autocomplete);
-    return result;
-  }, 'You must set "codaType" to ValueHintType.SelectList or ValueHintType.Reference when setting an "autocomplete" property.');
-const objectPackFormulaSchema = zodCompleteObject<Omit<ObjectPackFormula<any, any>, 'execute'>>({
-  ...commonPackFormulaSchema,
-  resultType: zodDiscriminant(Type.object),
-  // TODO(jonathan): See if we should really allow this. The SDK right now explicitly tolerates an undefined
-  // schema for objects, but that doesn't seem like a use case we actually want to support.
-  schema: z.union([genericObjectSchema, arrayPropertySchema]).optional(),
-});
-
-const formulaMetadataSchema = z
-  .union([numericPackFormulaSchema, stringPackFormulaSchema, booleanPackFormulaSchema, objectPackFormulaSchema])
-  .superRefine((data, context) => {
-    const parameters = data.parameters as ParamDefs;
-    const varargParameters = data.varargParameters || ([] as ParamDefs);
-    const paramNames = new Set();
-    for (const param of [...parameters, ...varargParameters]) {
-      if (paramNames.has(param.name)) {
-        context.addIssue({
-          code: z.ZodIssueCode.custom,
-          path: ['parameters'],
-          message: `Parameter names must be unique. Found duplicate name "${param.name}".`,
-        });
+  const objectPropertyUnionSchema = z
+    .union([
+      booleanPropertySchema,
+      numberPropertySchema,
+      stringPropertySchema,
+      arrayPropertySchema,
+      genericObjectSchema,
+    ])
+    .refine((schema: Schema) => {
+      if (sdkVersion && semver.satisfies(sdkVersion, '<=1.4.0')) {
```

Comment:
It would have been possible to use `schemaExtends` for this, but heuristic-y and awkward. It'd pull out `metadata.syncTables[].schema` to iterate through and check each one. Since array schemas don't put the `codaType` at the top level it would have to know to unwrap those schemas. Formulas and deeply-nested schemas likely wouldn't get checked for this to keep the code simple.

Moving away from `schemaExtends` toward having `sdkVersion` in the zod schemas themselves just felt like a simpler, cleaner, more future-proof approach.